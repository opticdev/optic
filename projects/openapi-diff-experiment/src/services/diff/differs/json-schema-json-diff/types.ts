import { BodyLocation, OpenAPIV3 } from '@useoptic/openapi-utilities';
import Ajv, { ErrorObject } from 'ajv';
import { ShapeDiffTypes } from '../../types';
import { JsonSchemaPatch } from './plugins/plugin-types';
import { JsonPatcher } from '../../../patch/incremental-json-patch/json-patcher';

export interface JsonSchemaJsonDiffer {
  ajvToDiff: (
    schema: OpenAPIV3.SchemaObject,
    schemaPath: string,
    diff: ErrorObject,
    example: any,
    location: BodyLocation
  ) => ShapeDiffTypes;
  diffToPatch: (
    diff: ShapeDiffTypes,
    patcher: JsonPatcher<OpenAPIV3.Document>
  ) => JsonSchemaPatch[];
  compare: (
    schema: OpenAPIV3.SchemaObject,
    to: any,
    location: BodyLocation,
    schemaPath: string,
    options: {
      collapseToFirstInstanceOfArrayDiffs: boolean;
    }
  ) => ShapeDiffTypes[];
  supportedKeywords: string[];
  opticAjv: Ajv;
  stripKeywordsNotSupported: (diffs: ErrorObject[]) => ErrorObject[];
}
