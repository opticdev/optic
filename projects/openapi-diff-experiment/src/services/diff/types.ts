/*

  Interface diffing traffic
   - produces a diff from the
   - debug one (just to get rolling)
     - HTTP Interaction
     - HAR
     - Postman?

   */
import { ApiTraffic } from '../traffic/types';
import { OpenAPIDiffingQuestions } from '../read/types';
import {
  ConceptualLocation,
  FieldLocation,
  OpenAPIV3,
} from '@useoptic/openapi-utilities';
import { JsonSchemaKnownKeyword } from './differs/json-schema-json-diff/plugins/plugin-types';
import { JsonSchemaJsonDiffer } from './differs/json-schema-json-diff/types';
import { JsonPath } from '@useoptic/openapi-io';

export interface IDiffService {
  compare(traffic: ApiTraffic): Promise<{ diffs: IDiff[]; errors: string[] }>;
  jsonSchemaDiffer: JsonSchemaJsonDiffer;
}

export type DiffServiceFactoryFunction = (
  spec: OpenAPIDiffingQuestions
) => IDiffService;

export enum DiffType {
  UnmatchedPath = 'UnmatchedPath',
  UnmatchedMethod = 'UnmatchedMethod',
  UnmatchedResponse = 'UnmatchedResponse',
  BodyAdditionalProperty = 'BodyAdditionalProperty',
  BodyMissingRequiredProperty = 'BodyMissingRequiredProperty',
  BodyUnmatchedType = 'BodyUnmatchedType',
  QueryAdditionalParameter = 'QueryAdditionalParameter',
}

export interface IDiff {
  type: DiffType;
}

export interface UnmatchedPath extends IDiff {
  type: DiffType.UnmatchedPath;
  path: string;
  method: OpenAPIV3.HttpMethods;
  closestMatch: string;
}

export interface UnmatchedMethod extends IDiff {
  type: DiffType.UnmatchedMethod;
  path: string;
  method: OpenAPIV3.HttpMethods;
}

export interface UnmatchedResponse extends IDiff {
  type: DiffType.UnmatchedResponse;
  path: string;
  method: OpenAPIV3.HttpMethods;
  statusCode: string;
}
////////////////////////////////////////////////////////////

export interface QueryAdditionalParameter extends IDiff {
  type: DiffType.QueryAdditionalParameter;
  path: string;
  name: string;
  method: OpenAPIV3.HttpMethods;
  example: any;
}

////////////////////////////////////////////////////////////

export interface BodyAdditionalProperty extends IDiff {
  type: DiffType.BodyAdditionalProperty;
  propertyExamplePath: JsonPath;
  parentObjectPath: JsonPath;
  key: string;
  example: any;
  propertyPath: JsonPath;
  keyword: JsonSchemaKnownKeyword.additionalProperties;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export interface BodyMissingRequiredProperty extends IDiff {
  type: DiffType.BodyMissingRequiredProperty;
  keyword: JsonSchemaKnownKeyword.required;
  parentObjectPath: JsonPath;
  propertyPath: JsonPath;
  key: string;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export interface BodyPropertyUnmatchedType extends IDiff {
  type: DiffType.BodyUnmatchedType;
  keyword: JsonSchemaKnownKeyword.type | JsonSchemaKnownKeyword.oneOf;
  propertyPath: JsonPath;
  key: string;
  example: any;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export type ShapeDiffTypes =
  | BodyMissingRequiredProperty
  | BodyAdditionalProperty
  | BodyPropertyUnmatchedType;

type DiffKinds =
  | UnmatchedMethod
  | UnmatchedPath
  | UnmatchedResponse
  | QueryAdditionalParameter
  | ShapeDiffTypes;

/*
  Diff Result Types
 */

export interface DiffResult {
  isMatch: boolean;
}
export interface MatchedWithContext<Context> extends DiffResult {
  isMatch: true;
  context: Context;
}
export interface Matched extends DiffResult {
  isMatch: true;
}

export interface Diff extends DiffResult {
  isMatch: false;
  diffs: DiffKinds[];
}

export interface DiffError extends DiffResult {
  isMatch: false;
  error: string;
}

export interface EitherDiffResult<Context = undefined> {
  context: Context;
  error: string;
  diffs: DiffKinds[];
  isMatch: boolean;
  isDiff: boolean;
}

export function EitherDiffResult<Context>(
  i: DiffResult
): EitherDiffResult<Context> {
  return {
    isDiff: !i.isMatch,
    isMatch: i.isMatch,
    context: i.hasOwnProperty('context')
      ? (i as MatchedWithContext<Context>).context
      : undefined,
    error: i.hasOwnProperty('error') ? (i as DiffError).error : undefined,
    diffs: i.hasOwnProperty('diffs') ? (i as Diff).diffs : [],
  };
}

export const DiffResult = {
  match: () => {
    return EitherDiffResult<undefined>({ isMatch: true } as Matched);
  },
  matchWithContext: <Context>(context: Context) => {
    return EitherDiffResult<Context>({
      isMatch: true,
      context,
    } as MatchedWithContext<Context>);
  },
  diff: (diffs: DiffKinds[]) => {
    return EitherDiffResult<undefined>({
      isMatch: false,
      diffs,
    } as Diff);
  },
  error: (error: string) => {
    return EitherDiffResult<undefined>({ isMatch: false, error } as DiffError);
  },
};

export function isSchemaDiff(i: IDiff): boolean {
  return (
    i.type === DiffType.BodyAdditionalProperty ||
    i.type === DiffType.BodyUnmatchedType ||
    i.type === DiffType.BodyMissingRequiredProperty
  );
}
