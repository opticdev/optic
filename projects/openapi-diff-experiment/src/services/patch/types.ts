import { OpenAPIV3 } from '@useoptic/openapi-utilities';
import { ISpecReader } from '../read/types';
import {
  IPatchGroup,
  JsonPatcher,
  PatchesToSave,
} from './incremental-json-patch/json-patcher';
import { ApiTraffic } from '../traffic/types';
import { ShapeDiffTypes } from '../diff/types';
import { JsonSchemaPatch } from '../diff/differs/json-schema-json-diff/plugins/plugin-types';

/*
  Patching a specification has two steps
  - mutating the JSON-like object, by exposing domain-relevant helpers
  - updating the JSON-like objects across the file system to match w/ what we have in-memory via a reconcile pattern. Json-like * sourcemap
    essentially redraw the JSON across the file system with our changes
 */

/*
  During WIP Phase, we will not be reconciling the file system -- instead we will simple, be flushing via stringify methods.
  It's destructive, but not the part to optimize first
 */

export interface IPatchOpenAPI {
  reset: () => void;
  listPatches: () => IPatchGroup[];
  save: (options: { dryRun: boolean }) => Promise<IFilePatch>;
  fork: () => Promise<IPatchOpenAPI>;
  forkedPatcher(): JsonPatcher<OpenAPIV3.Document>;
  init: {
    operation: (
      pathPattern: string,
      method: OpenAPIV3.HttpMethods,
      example: ApiTraffic
    ) => void;
    queryParameter: (
      method: OpenAPIV3.HttpMethods,
      path: string,
      name: string,
      exampleValue: string | boolean | number
    ) => void;
    response: (
      pathPattern: string,
      method: OpenAPIV3.HttpMethods,
      example: ApiTraffic
    ) => void;
    responseBody: (
      pathPattern: string,
      method: OpenAPIV3.HttpMethods,
      statusCodeMatcher: string,
      contentType: string,
      example: ApiTraffic
    ) => void;
  };

  patch: {
    property: (jsonSchemaPatch: JsonSchemaPatch) => void;
  };
}

export interface IPatchOpenAPIReconciler {
  patchesToFileMutations(
    patchesToSave: PatchesToSave<OpenAPIV3.Document>
  ): Promise<IFilePatch>;
}

export type IPatchOpenAPIReconcilerFactory = (
  reader: ISpecReader
) => IPatchOpenAPIReconciler;

export type IFilePatch = {
  files: { path: string; previousContents: string; newContents: string }[];
};
