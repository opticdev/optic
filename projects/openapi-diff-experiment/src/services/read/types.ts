/*

  OpenAPI projections for diffs
   - assemble all the information required to diff
   - retrain pointers to ast nodes

   */
import {
  BodyLocation,
  OpenAPIV3,
  QueryParameterLocation,
} from '@useoptic/openapi-utilities';
import { IFilePatch } from '../patch/types';
import { JsonSchemaSourcemap } from '@useoptic/openapi-io';

export interface OpenAPIDiffingQuestions {
  paths(): string[];
  operations(): {
    path: string;
    method: OpenAPIV3.HttpMethods;
    jsonPath: string;
  }[];
  responsesForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): ResponseMatchType[];

  requestBodiesForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): RequestBodyMatchType[];

  queryParametersForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): QueryParameterType[];
}

export interface ISpecReader {
  questions(): Promise<OpenAPIDiffingQuestions>;
  sourcemap(): Promise<JsonSchemaSourcemap>;
  flattenedSpecification(): Promise<OpenAPIV3.Document>;
  didLoad(): Promise<DidLoadStatus>;
  save(patch: IFilePatch): Promise<void>;
  describeLocation(): string;
  rootFile(): string;
  reload(): Promise<void>;
  mode: 'simulated' | 'filesystem';
}

export type FilePathsWithChanges = string[];
export type DidLoadStatus =
  | { success: true; durationMillis: number }
  | { success: false; error: string; durationMillis: number };

export const openApiDiffingQuestionsTestingStub: OpenAPIDiffingQuestions = {
  operations() {
    return [];
  },
  paths(): string[] {
    return [];
  },
  requestBodiesForOperation(method: OpenAPIV3.HttpMethods, path: string) {
    return [];
  },
  responsesForOperation(method: OpenAPIV3.HttpMethods, path: string) {
    return [];
  },
  queryParametersForOperation(method: OpenAPIV3.HttpMethods, path: string) {
    return [];
  },
};

// Result Types
export type ResponseMatchType = {
  statusCodeMatcher: string;
  contentTypes: {
    contentType: string;
    schema?: OpenAPIV3.SchemaObject;
    location: BodyLocation;
    jsonPath: string;
  }[];
};

export type RequestBodyMatchType = {
  contentType: string;
  schema?: OpenAPIV3.SchemaObject;
  location: BodyLocation;
  jsonPath: string;
};

export type QueryParameterType = {
  name: string;
  schema?: OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject;
  location: QueryParameterLocation;
  jsonPath: string;
  required: boolean;
};
