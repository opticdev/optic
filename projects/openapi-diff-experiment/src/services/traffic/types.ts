import events from 'events';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
/*

  Interface for traffic, expose what the differ cares about via interfaces
   - different formats can have different implementations
   - debug one (just to get rolling)
     - HTTP Interaction
     - HAR
     - Postman?

   */
export interface ApiTraffic {
  method: OpenAPIV3.HttpMethods;
  path: string;
  queryString: string;
  response: {
    statusCode: string;
    body: {
      contentType?: string;
      jsonBodyString?: string;
    };
  };
  requestBody?: {
    contentType?: string;
    jsonBodyString?: string;
  };
}

/*

  Interface for traffic, should get traffic (somehow), and emit each interaction into our queue
   - each implementation should be self contained starting / parsing / stopping traffic collection
   - known types
     - file
     - in-memory (for debugging)
     - proxy
     - postman
     - sniff
     - chrome

   */

export interface ITrafficSource {
  start(): Promise<void>;
  stop(): Promise<void>;
  emitTraffic(traffic: ApiTraffic): Promise<void>;
  on(event: 'traffic', listener: (example: ApiTraffic) => void): this;
}

export class TrafficSource
  extends events.EventEmitter
  implements ITrafficSource
{
  emitTraffic(traffic: ApiTraffic): Promise<void> {
    this.emit('traffic', traffic);
    return Promise.resolve();
  }

  start(): Promise<void> {
    return Promise.resolve(undefined);
  }

  stop(): Promise<void> {
    return Promise.resolve(undefined);
  }
}
