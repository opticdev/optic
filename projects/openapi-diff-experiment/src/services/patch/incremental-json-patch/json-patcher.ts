import jsonPatch, { Operation } from 'fast-json-patch';
import { copyObject } from '../../../utils/debug_waitFor';
import { isObject } from '../../../utils/is-object';
import niceTry from 'nice-try';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { isExtension } from '../../diff/differs/json-schema-json-diff/transition-assumptions';
import invariant from 'ts-invariant';

export interface JsonPatcher<G> {
  fork: () => JsonPatcher<G>;
  currentPatches: () => IPatchGroup[];
  currentPatchesRelativeTo: (basePointer: string) => IPatchGroup[];
  applyPatch: (patch: IPatchGroup) => void;
  helper: {
    getPath: (basePath: string, path: string[]) => any;
    get: (path: string) => any;
    removeKeysNotAllowedAt: (
      basePath: string,
      allowedKeys: string[],
      reason: string
    ) => void;
  };
  apply: (
    intent: string,
    patches: Operation[],
    swallowErrors?: boolean
  ) => {
    error?: string;
    patch?: { patches: Operation[]; intent: string };
    success: boolean;
  };
  reset: () => void;
  currentDocument: (log?: boolean) => G;
  patchesToSave: () => PatchesToSave<G>;
}

export function jsonPatcher<G>(
  original: G,
  initialPatches: IPatchGroup[] = []
): JsonPatcher<G> {
  let currentPatches: IPatchGroup[] = initialPatches;

  const currentDocument = (log: boolean = false) => {
    return currentPatches.reduce((original, patch) => {
      const newDoc = jsonPatch.applyPatch(
        copyObject(original),
        copyObject(patch.patches),
        true,
        false
      ).newDocument;
      if (log) {
        console.log(`apply patch: ${patch.intent}`);
        console.log(
          `with operations: ${JSON.stringify(patch.patches, null, 3)}`
        );
        console.log(`and result: ${JSON.stringify(newDoc, null, 3)}`);
      }
      return newDoc as G;
    }, copyObject(original) as G);
  };

  const apply = (
    intent: string,
    patches: Operation[],
    swallowErrors: boolean = false
  ) => {
    try {
      jsonPatch.applyPatch(
        copyObject(currentDocument()),
        copyObject(patches),
        true,
        false
      );
      const patch = { intent, patches };
      currentPatches.push(patch);
      return { success: true, patch };
    } catch (e: any) {
      if (swallowErrors) return { success: false, error: e.message };
      invariant(false, 'could not apply json patch ' + e.message);
    }
  };

  return {
    fork: () => {
      return jsonPatcher(original, [
        ...(copyObject(currentPatches) as IPatchGroup[]),
      ]);
    },
    currentPatches: () => {
      return currentPatches;
    },
    currentPatchesRelativeTo: (basePointer: string) => {
      const allPatches: IPatchGroup[] = copyObject(currentPatches || []);
      return allPatches.map((p) => {
        const { patches } = p;
        const updatedPatches = patches.map((operation) => {
          return {
            ...operation,
            path: jsonPointerHelpers.append(
              basePointer,
              ...jsonPointerHelpers.decode(operation.path)
            ),
          };
        });
        return { ...p, patches: updatedPatches };
      });
    },
    currentDocument,
    patchesToSave: (): PatchesToSave<G> => {
      return { patches: currentPatches, document: currentDocument() };
    },
    apply,
    applyPatch: (patch: IPatchGroup) => {
      apply(patch.intent, patch.patches);
    },
    reset: () => {
      currentPatches = [];
    },
    helper: {
      get: (path: string) => {
        return niceTry(() =>
          jsonPatch.getValueByPointer(currentDocument(), path)
        );
      },
      removeKeysNotAllowedAt: (
        basePath: string,
        allowedKeys: string[],
        reason: string
      ) => {
        const current = currentDocument();
        const parent = niceTry(() =>
          jsonPatch.getValueByPointer(current, basePath)
        );

        if (parent && isObject(parent)) {
          Object.keys(parent).forEach((key) => {
            if (!allowedKeys.includes(key) || isExtension(key)) {
              const keypath = jsonPointerHelpers.append(basePath, key);
              apply(`remove ${key} after ${reason}`, [
                {
                  op: 'remove',
                  path: keypath,
                },
              ]);
            }
          });
        }
      },
      getPath: (basePath: string, paths: string[]) => {
        const myPath = jsonPointerHelpers.append(basePath, ...paths);
        return niceTry(() =>
          jsonPatch.getValueByPointer(currentDocument(), myPath)
        );
      },
    },
  };
}

export interface IPatchGroup {
  intent: string; // human readable
  patches: Operation[];
}

export type PatchesToSave<G> = { patches: IPatchGroup[]; document: G };
