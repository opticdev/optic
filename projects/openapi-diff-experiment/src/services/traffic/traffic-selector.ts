import { ApiTraffic } from './types';
import { ConceptualLocation } from '@useoptic/openapi-utilities';
import invariant from 'ts-invariant';

export function trafficSelector(traffic: ApiTraffic) {
  return {
    bodyJsonForLocation: (location: ConceptualLocation) => {
      if ('inResponse' in location && 'body' in location.inResponse) {
        const contentType = location.inResponse.body.contentType;

        invariant(
          contentType === traffic.response.body.contentType,
          'content type must match when selecting examples from traffic'
        );

        invariant(
          typeof traffic.response.body.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );

        return JSON.parse(traffic.response.body.jsonBodyString!);
      }

      if ('inRequest' in location && 'body' in location.inRequest) {
        const { contentType } = location.inRequest.body;
        invariant(
          contentType === traffic.requestBody?.contentType,
          'content type must match when selecting examples from traffic'
        );
        invariant(
          typeof traffic.requestBody.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );
        return JSON.parse(traffic.requestBody.jsonBodyString!);
      }

      invariant(false, 'can not select example body from traffic');
    },
  };
}
