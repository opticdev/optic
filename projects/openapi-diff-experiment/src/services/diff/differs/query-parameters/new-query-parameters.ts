import {
  IPatchGroup,
  JsonPatcher,
} from '../../../patch/incremental-json-patch/json-patcher';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
import { JsonPath } from '@useoptic/openapi-io';
import { ApiTraffic } from '../../../traffic/types';
import { parseQueryStringToMap } from './optic-query-string-parser';
import { opticJsonSchemaDiffer } from '../json-schema-json-diff';
import { streamingJsonSchemaBuilder } from '../json-schema-json-diff/json-builder/streaming-json-schema-builder';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';

export function newAddAllQueryParameters(
  patcher: JsonPatcher<OpenAPIV3.Document>,
  method: OpenAPIV3.HttpMethods,
  path: string,
  example: ApiTraffic
) {
  const queryParams = parseQueryStringToMap(example.queryString);

  const newParameters: OpenAPIV3.ParameterObject[] = Object.entries(
    queryParams
  ).map((queryObservation) => {
    const [name, value] = queryObservation;

    // start all query params as false
    return newQueryParameter(name, value as any, false);
  });

  if (newParameters.length > 0) {
    const current = patcher.currentDocument();
    const parametersPath = jsonPointerHelpers.compile([
      'paths',
      path,
      method,
      'parameters',
    ]);

    //ensure parameters array exists
    const parametersArray = jsonPointerHelpers.tryGet(current, parametersPath);
    if (
      !parametersArray.match ||
      (parametersArray.match && !Array.isArray(parametersArray.value))
    ) {
      patcher.apply(`add parameters array for ${method} ${path}`, [
        {
          op: 'add',
          path: parametersPath,
          value: [],
        },
      ]);
    }

    newParameters.forEach((param) => {
      const conflict = Boolean(
        parametersArray.match &&
          parametersArray.value.find(
            (i: OpenAPIV3.ParameterObject) =>
              i.name === param.name && i.in === 'query'
          )
      );
      if (!conflict) {
        patcher.apply(`add parameters array for ${method} ${path}`, [
          {
            op: 'add',
            path: jsonPointerHelpers.append(parametersPath, '-'),
            value: param,
          },
        ]);
      }
    });
  }
}

export function newQueryParameter(
  name: string,
  // will eventually include arrays and objects
  value: string | boolean | number,
  required: boolean
): OpenAPIV3.ParameterObject {
  const schema = streamingJsonSchemaBuilder(opticJsonSchemaDiffer(), value);
  return {
    name,
    required,
    schema,
    in: 'query',
  };
}

export function patchAdditionalQueryParameter(
  patcher: JsonPatcher<OpenAPIV3.Document>,
  method: OpenAPIV3.HttpMethods,
  path: string,
  name: string,
  exampleValue: string | boolean | number
) {
  const current = patcher.currentDocument();
  const parametersPath = jsonPointerHelpers.compile([
    'paths',
    path,
    method,
    'parameters',
  ]);

  const param = newQueryParameter(name, exampleValue, false);

  //ensure parameters array exists
  const parametersArray = jsonPointerHelpers.tryGet(current, parametersPath);
  if (
    !parametersArray.match ||
    (parametersArray.match && !Array.isArray(parametersArray.value))
  ) {
    patcher.apply(`add parameters array for ${method} ${path}`, [
      {
        op: 'add',
        path: parametersPath,
        value: [],
      },
    ]);
  }

  const conflict = Boolean(
    parametersArray.match &&
      parametersArray.value.find(
        (i: OpenAPIV3.ParameterObject) => i.name === name && i.in === 'query'
      )
  );
  if (!conflict) {
    patcher.apply(`add parameters array for ${method} ${path}`, [
      {
        op: 'add',
        path: jsonPointerHelpers.append(parametersPath, '-'),
        value: param,
      },
    ]);
  }
}
