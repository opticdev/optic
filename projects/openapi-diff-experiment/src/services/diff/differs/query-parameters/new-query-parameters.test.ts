import { ApiTraffic } from '../../../traffic/types';
import { OpenAPIV3 } from 'openapi-types';
import { newAddAllQueryParameters } from './new-query-parameters';
import { jsonPatcher } from '../../../patch/incremental-json-patch/json-patcher';

const makeRequestWithQueryString = (query: string) => {
  const traffic: ApiTraffic = {
    path: '/example',
    method: OpenAPIV3.HttpMethods.GET,
    queryString: query,
    response: {
      statusCode: '200',
      body: {},
    },
  };
  return traffic;
};

const emptySpec: OpenAPIV3.Document = {
  openapi: '3.0.3',
  info: { title: 'empty', version: '0.0.0' },
  paths: {
    '/example': {
      get: {
        responses: {},
      },
    },
  },
};
const emptySpecWithAParameterAlreadyDefined: OpenAPIV3.Document = {
  openapi: '3.0.3',
  info: { title: 'empty', version: '0.0.0' },
  paths: {
    '/example': {
      get: {
        parameters: [
          {
            in: 'query',
            name: 'pre-existing',
            required: true,
          },
        ],
        responses: {},
      },
    },
  },
};

describe('query parameters', () => {
  describe('init', () => {
    it('can document all query parameters for traffic', () => {
      const patch = jsonPatcher(emptySpec);
      const request = makeRequestWithQueryString('search=Hello&limit=30');
      newAddAllQueryParameters(
        patch,
        OpenAPIV3.HttpMethods.GET,
        '/example',
        request
      );

      expect(patch.currentDocument()).toMatchSnapshot();
    });

    it('can document all query parameters for traffic when 1 query param already in spec', () => {
      const patch = jsonPatcher(emptySpecWithAParameterAlreadyDefined);
      const request = makeRequestWithQueryString('search=Hello&limit=30');
      newAddAllQueryParameters(
        patch,
        OpenAPIV3.HttpMethods.GET,
        '/example',
        request
      );

      expect(patch.currentDocument()).toMatchSnapshot();
    });

    it('empty query string will document empty', () => {
      const patch = jsonPatcher(emptySpec);
      const request = makeRequestWithQueryString('');
      newAddAllQueryParameters(
        patch,
        OpenAPIV3.HttpMethods.GET,
        '/example',
        request
      );

      expect(patch.currentDocument()).toMatchSnapshot();
    });
  });
});
