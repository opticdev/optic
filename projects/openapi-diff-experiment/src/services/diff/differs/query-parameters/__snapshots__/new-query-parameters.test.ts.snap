// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query parameters init can document all query parameters for traffic 1`] = `
Object {
  "info": Object {
    "title": "empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": Object {
              "type": "number",
            },
          },
          Object {
            "in": "query",
            "name": "search",
            "required": false,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {},
      },
    },
  },
}
`;

exports[`query parameters init can document all query parameters for traffic when 1 query param already in spec 1`] = `
Object {
  "info": Object {
    "title": "empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "query",
            "name": "pre-existing",
            "required": true,
          },
          Object {
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": Object {
              "type": "number",
            },
          },
          Object {
            "in": "query",
            "name": "search",
            "required": false,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {},
      },
    },
  },
}
`;

exports[`query parameters init empty query string will document empty 1`] = `
Object {
  "info": Object {
    "title": "empty",
    "version": "0.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "responses": Object {},
      },
    },
  },
}
`;
